pkg/planner/core/operator/physicalop/physical_indexlookup_reader.go

```diff
@@ -205,14 +222,52 @@ func (p *PhysicalIndexLookUpReader) ExplainInfo() string {
 }
 
 // Init initializes PhysicalIndexLookUpReader.
-func (p PhysicalIndexLookUpReader) Init(ctx base.PlanContext, offset int) *PhysicalIndexLookUpReader {
+func (p PhysicalIndexLookUpReader) Init(ctx base.PlanContext, offset int, tryPushDownIndexLookUp bool) *PhysicalIndexLookUpReader {
 	p.BasePhysicalPlan = NewBasePhysicalPlan(ctx, plancodec.TypeIndexLookUp, &p, offset)
-	p.TablePlans = FlattenPushDownPlan(p.TablePlan)
-	p.IndexPlans = FlattenPushDownPlan(p.IndexPlan)
 	p.SetSchema(p.TablePlan.Schema())
+	p.SetStats(p.TablePlan.StatsInfo())
+	if tryPushDownIndexLookUp {
+		p.tryPushDownLookUp(ctx)
+	}
+	p.TablePlans = FlattenListOrTiFlashPushDownPlan(p.TablePlan)
+	p.IndexPlans, p.IndexPlansUnNatureOrders = FlattenTreePushDownPlan(p.IndexPlan)
 	return &p
 }
 
+// tryPushDownLookUp tries to push down the index lookup to TiKV.
+func (p *PhysicalIndexLookUpReader) tryPushDownLookUp(ctx base.PlanContext) {
+	intest.Assert(!p.IndexLookUpPushDown)
+	if p.KeepOrder {
+		// If keep order is required, we cannot push down the index lookup.
+		ctx.GetSessionVars().StmtCtx.SetHintWarning("hint INDEX_LOOKUP_PUSHDOWN is inapplicable, keep order is not supported.")
+		return
+	}
+
+	indexLookUpPlan, err := buildPushDownIndexLookUpPlan(ctx, p.IndexPlan, p.TablePlan)
+	if err != nil {
+		// This should not happen, but if it happens, we just log a warning and continue to use the original plan.
+		intest.AssertNoError(err)
+		logutil.BgLogger().Warn("try to push down index lookup failed", zap.Error(err))
+		return
+	}
+	p.IndexPlan = indexLookUpPlan
+	// Currently, it's hard to estimate how many rows can be looked up locally when push-down.
+	// So we just use the row count as 0 of tablePlan in TiDB side which displays all lookup
+	// can be performed in the TiKV side.
+	scalePlanStatusInfoAsZeroRecursively(ctx.GetSessionVars(), p.TablePlan)
+	// The status info of IndexLookupReader should be the same as indexPlan in the push-down mode if
+	// all lookup can be performed in the TiKV side.
+	p.SetStats(p.IndexPlan.StatsInfo())
+	p.IndexLookUpPushDown = true
+}
+
+func scalePlanStatusInfoAsZeroRecursively(vars *variable.SessionVars, p base.PhysicalPlan) {
```

Comment:
```suggestion
func resetRowCountAsZeroRecursively(vars *variable.SessionVars, p base.PhysicalPlan) {
```
Though we may not need this function.
You can change its name first. `Stats` is short for `statistics`, not `status`.